### PR TITLE
[Bugfix] Fix RealESRGAN fp16 dtype mismatch and cache keying

### PR DESCRIPTION
**Pull request by jy-song-hub**
## Motivation

RealESRGAN upscaling supports  `half_precision=True` , but the previous implementation could (1) crash due to fp16 weight / fp32 input dtype mismatches and (2) incorrectly reuse cached models across fp16/fp32 (and across fp16 requests that fall back to fp32), because the cache key did not include the effective dtype. This shows up in mixed-configuration workloads and can cause incorrect behavior or hard failures. 

## Modifications
- Cast the input tensor to the model’s parameter dtype inside  `UpscalerModel.upscale()`  so fp16 weights always receive fp16 inputs (prevents  `Float/Half`  mismatch crashes). 
- Key the global RealESRGAN model cache by  `(resolved_path, device, dtype)`  (dtype is the effective inference dtype) to avoid fp16/fp32 collisions and avoid poisoning the cache when fp16 is requested but execution falls back to fp32. 
- Convert output to float32 on CPU before NumPy conversion ( `out.cpu().float().numpy()` ) to avoid an extra GPU cast+copy for fp16 outputs.  
- Refine fp16 enablement on CUDA by gating on minimum compute capability; fall back to fp32 when capability is unknown/too old. Keep fp16 enabled on ROCm/MPS/MUSA/NPU where supported. 

## Accuracy Tests
This change does not alter model math beyond matching the input dtype to the already-selected inference dtype. See comments for the unit test we used.   

## Speed Tests and Profiling
Not applicable. No performance‑critical kernels were changed; cache lookups remain O(1). 
- fp32 path: no meaningful change. 
- fp16 path: prevents dtype-mismatch crashes; minor overhead reduction from doing the fp16→fp32 conversion on CPU before NumPy.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #26783756098](https://github.com/sgl-project/sglang/actions/runs/26783756098)
Latest PR Test (Extra): :x: [Run #26783755873](https://github.com/sgl-project/sglang/actions/runs/26783755873)

### CODE REVIEW

File: python/sglang/multimodal_gen/test/unit/test_postprocess_cache.py [hunk 1]
```diff
@@ -0,0 +1,277 @@
+"""
+Test for postprocess model cache keying bug.
```
Comment (yyy1000) [inline python/sglang/multimodal_gen/test/unit/test_postprocess_cache.py]: Thank you for the PR! I’m wondering whether we should keep this test file, as it doesn’t cover core functionality and may unnecessarily bloat the codebase.

Comment (jy-song-hub) [reply python/sglang/multimodal_gen/test/unit/test_postprocess_cache.py]: Thanks for the feedback! I added the test file mainly to reproduce/regression-test the original cache keying issue. I’m totally OK removing it if you prefer to keep the codebase lean—I can instead summarize the repro steps and expected behavior in the PR description
Do you have a preference on which approach is better?

Comment (yyy1000) [reply python/sglang/multimodal_gen/test/unit/test_postprocess_cache.py]: Sure, I prefer not adding the test code but would like others to take a look. Otherwise it looks good to me.

Comment (jy-song-hub) [reply python/sglang/multimodal_gen/test/unit/test_postprocess_cache.py]: Thanks! Any suggestions here~ @mickqian

Comment (mickqian) [reply python/sglang/multimodal_gen/test/unit/test_postprocess_cache.py]: agree, we could remove it

Comment (jy-song-hub) [reply python/sglang/multimodal_gen/test/unit/test_postprocess_cache.py]: Unit test removed. Thank you for your suggestions! @yyy1000 @mickqian

**Comment by jy-song-hub** [comment]
We ran unit tests to validate the correctness of this PR. The test code is not included to avoid increasing codebase size;  code is provided below for reference.

```
import unittest
from unittest.mock import patch

import numpy as np
import torch

# Import the modules to access their caches
from sglang.multimodal_gen.runtime.postprocess import (
    realesrgan_upscaler,
    rife_interpolator,
)


class TestPostprocessCacheKeying(unittest.TestCase):
    """Test cache keying by precision and device."""

    def setUp(self):
        """Clear module caches before each test."""
        realesrgan_upscaler._MODEL_CACHE.clear()
        rife_interpolator._MODEL_CACHE.clear()

    def _create_dummy_realesrgan_state_dict(self):
        """Create a minimal state dict that RealESRGAN's _build_net_from_state_dict will accept."""
        # SRVGGNetCompact detection looks for "body.0.weight" etc.
        # Create a dummy that looks like SRVGGNetCompact with num_feat=64, num_conv=16, upscale=4
        state_dict = {}
        # First conv: in_ch=3, out_ch=num_feat=64
        state_dict["body.0.weight"] = torch.randn(64, 3, 3, 3)
        state_dict["body.0.bias"] = torch.randn(64)
        # First activation (PReLU) - not a weight, but may be present
        state_dict["body.1.weight"] = torch.randn(64)  # PReLU alpha

        # Body convs: 16 conv layers, each followed by activation
        for i in range(16):
            conv_idx = 2 + i * 2
            act_idx = conv_idx + 1
            state_dict[f"body.{conv_idx}.weight"] = torch.randn(64, 64, 3, 3)
            state_dict[f"body.{conv_idx}.bias"] = torch.randn(64)
            state_dict[f"body.{act_idx}.weight"] = torch.randn(64)

        # Last conv: maps to out_ch * upscale^2 = 3 * 16 = 48
        state_dict["body.34.weight"] = torch.randn(48, 64, 3, 3)
        state_dict["body.34.bias"] = torch.randn(48)

        return state_dict

    def _create_dummy_rife_state_dict(self):
        """Create a minimal state dict that RIFE's load_model will accept."""
        # RIFE loads flownet.pkl which contains the full IFNet state dict
        # We'll create a minimal one with the expected keys
        state_dict = {}
        # A few representative keys from IFNet
        state_dict["block0.conv0.0.weight"] = torch.randn(96, 7 + 8, 3, 3)
        state_dict["block0.conv0.0.bias"] = torch.randn(96)
        state_dict["block0.conv0.1.weight"] = torch.randn(96)
        state_dict["block0.conv0.1.bias"] = torch.randn(96)
        state_dict["encode.cnn0.weight"] = torch.randn(16, 3, 3, 3)
        state_dict["encode.cnn0.bias"] = torch.randn(16)
        return state_dict

    @patch("torch.load")
    @patch(
        "sglang.multimodal_gen.runtime.postprocess.realesrgan_upscaler._resolve_model_path"
    )
    @patch(
        "sglang.multimodal_gen.runtime.platforms.current_platform.get_local_torch_device"
    )
    @patch("torch.nn.modules.module.Module.to", autospec=True)
    @patch.object(realesrgan_upscaler.UpscalerModel, "upscale")
    def test_realesrgan_precision_cache_bug(
        self,
        mock_upscale,
        mock_module_to,
        mock_get_device,
        mock_resolve_path,
        mock_torch_load,
    ):
        """Test that RealESRGAN cache ignores half_precision setting."""
        # Setup mocks
        mock_module_to.side_effect = lambda module, *args, **kwargs: module
        mock_get_device.return_value = torch.device("cuda:0")
        mock_resolve_path.return_value = "/dummy/model.pth"
        mock_torch_load.return_value = self._create_dummy_realesrgan_state_dict()
        # Mock the upscale method to avoid actual computation
        mock_upscale.return_value = np.zeros((128, 128, 3), dtype=np.uint8)

        # 1. Load with half_precision=True
        from sglang.multimodal_gen.runtime.postprocess.realesrgan_upscaler import (
            ImageUpscaler,
        )

        upscaler1 = ImageUpscaler(model_path="/dummy/model.pth", half_precision=True)

        # Trigger model loading (but not actual forward pass)
        dummy_frame = np.random.randint(0, 256, size=(64, 64, 3), dtype=np.uint8)
        _ = upscaler1.upscale([dummy_frame])

        # Check cache state
        self.assertEqual(len(realesrgan_upscaler._MODEL_CACHE), 1)
        cached_model1 = list(realesrgan_upscaler._MODEL_CACHE.values())[0]
        # Model should be fp16
        self.assertEqual(next(cached_model1.net.parameters()).dtype, torch.float16)

        # 2. Load with half_precision=False (same model path)
        upscaler2 = ImageUpscaler(model_path="/dummy/model.pth", half_precision=False)
        _ = upscaler2.upscale([dummy_frame])

        # WITH FIX: Cache should have 2 entries (fp16 and fp32 separate)
        self.assertEqual(len(realesrgan_upscaler._MODEL_CACHE), 2)

        # Check dtypes: expect one fp16 and one fp32 model
        dtypes = {
            next(model.net.parameters()).dtype
            for model in realesrgan_upscaler._MODEL_CACHE.values()
        }
        self.assertEqual(dtypes, {torch.float16, torch.float32})

    @patch("torch.load")
    @patch(
        "sglang.multimodal_gen.runtime.postprocess.realesrgan_upscaler._resolve_model_path"
    )
    @patch(
        "sglang.multimodal_gen.runtime.platforms.current_platform.get_local_torch_device"
    )
    @patch("torch.nn.modules.module.Module.to", autospec=True)
    @patch.object(realesrgan_upscaler.UpscalerModel, "upscale")
    def test_realesrgan_device_cache_bug(
        self,
        mock_upscale,
        mock_module_to,
        mock_get_device,
        mock_resolve_path,
        mock_torch_load,
    ):
        """Test that RealESRGAN cache ignores device setting."""
        mock_module_to.side_effect = lambda module, *args, **kwargs: module
        # Mock 1: CUDA device
        mock_get_device.side_effect = [torch.device("cuda:0"), torch.device("cpu")]
        mock_resolve_path.return_value = "/dummy/model.pth"
        mock_torch_load.return_value = self._create_dummy_realesrgan_state_dict()
        mock_upscale.return_value = np.zeros((128, 128, 3), dtype=np.uint8)

        # Load on CUDA
        from sglang.multimodal_gen.runtime.postprocess.realesrgan_upscaler import (
            ImageUpscaler,
        )

        upscaler1 = ImageUpscaler(model_path="/dummy/model.pth")
        dummy_frame = np.random.randint(0, 256, size=(64, 64, 3), dtype=np.uint8)
        _ = upscaler1.upscale([dummy_frame])

        # Check cache size and model load behavior
        self.assertEqual(len(realesrgan_upscaler._MODEL_CACHE), 1)
        self.assertEqual(mock_module_to.call_count, 1)

        # Load on CPU (same model path)
        upscaler2 = ImageUpscaler(model_path="/dummy/model.pth")
        _ = upscaler2.upscale([dummy_frame])

        # WITH FIX: Cache should now have 2 entries (separate device keys)
        self.assertEqual(len(realesrgan_upscaler._MODEL_CACHE), 2)
        # A new load should trigger another .to() call
        self.assertEqual(mock_module_to.call_count, 2)

    @patch("torch.load")
    @patch("sglang.multimodal_gen.runtime.postprocess.rife_interpolator.os.path.isfile")
    @patch(
        "sglang.multimodal_gen.runtime.utils.hf_diffusers_utils.maybe_download_model"
    )
    @patch(
        "sglang.multimodal_gen.runtime.platforms.current_platform.get_local_torch_device"
    )
    @patch("torch.nn.modules.module.Module.to", autospec=True)
    @patch.object(rife_interpolator.Model, "inference")
    def test_rife_device_cache_bug(
        self,
        mock_inference,
        mock_module_to,
        mock_get_device,
        mock_maybe_download,
        mock_isfile,
        mock_torch_load,
    ):
        """Test that RIFE cache ignores device setting."""
        mock_module_to.side_effect = lambda module, *args, **kwargs: module
        mock_isfile.return_value = True
        # Mock 1: CUDA device
        mock_get_device.side_effect = [torch.device("cuda:0"), torch.device("cpu")]
        mock_maybe_download.return_value = "/dummy/rife/model"
        mock_torch_load.return_value = self._create_dummy_rife_state_dict()
        # Mock inference to avoid actual computation
        mock_inference.return_value = torch.randn(1, 3, 64, 64)

        # Load on CUDA
        from sglang.multimodal_gen.runtime.postprocess.rife_interpolator import (
            FrameInterpolator,
        )

        interpolator1 = FrameInterpolator(model_path="/dummy/rife/model")

        dummy_frame = np.random.randint(0, 256, size=(64, 64, 3), dtype=np.uint8)
        _ = interpolator1.interpolate([dummy_frame, dummy_frame], exp=1)

        # Check cache size and model load behavior
        self.assertEqual(len(rife_interpolator._MODEL_CACHE), 1)
        self.assertEqual(mock_module_to.call_count, 1)

        # Load on CPU (same model path)
        interpolator2 = FrameInterpolator(model_path="/dummy/rife/model")
        _ = interpolator2.interpolate([dummy_frame, dummy_frame], exp=1)

        # WITH FIX: Cache should now have 2 entries (separate device keys)
        self.assertEqual(len(rife_interpolator._MODEL_CACHE), 2)
        # A new load should trigger another .to() call
        self.assertEqual(mock_module_to.call_count, 2)

    @patch("torch.load")
    @patch(
        "sglang.multimodal_gen.runtime.postprocess.realesrgan_upscaler._resolve_model_path"
    )
    @patch(
        "sglang.multimodal_gen.runtime.platforms.current_platform.get_local_torch_device"
    )
    @patch("torch.nn.modules.module.Module.to", autospec=True)
    @patch.object(realesrgan_upscaler.UpscalerModel, "upscale")
    def test_cache_independence_after_fix(
        self,
        mock_upscale,
        mock_module_to,
        mock_get_device,
        mock_resolve_path,
        mock_torch_load,
    ):
        """Test that after fix, different configs create separate cache entries."""
        mock_module_to.side_effect = lambda module, *args, **kwargs: module
        # This test will pass after the fix is implemented
        mock_resolve_path.return_value = "/dummy/model.pth"
        mock_torch_load.return_value = self._create_dummy_realesrgan_state_dict()
        mock_upscale.return_value = np.zeros((128, 128, 3), dtype=np.uint8)

        # 1. CUDA, fp16
        mock_get_device.side_effect = [
            torch.device("cuda:0"),
            torch.device("cuda:0"),
            torch.device("cpu"),
        ]
        from sglang.multimodal_gen.runtime.postprocess.realesrgan_upscaler import (
            ImageUpscaler,
        )

        upscaler1 = ImageUpscaler(model_path="/dummy/model.pth", half_precision=True)
        dummy_frame = np.random.randint(0, 256, size=(64, 64, 3), dtype=np.uint8)
        _ = upscaler1.upscale([dummy_frame])

        # 2. CUDA, fp32
        upscaler2 = ImageUpscaler(model_path="/dummy/model.pth", half_precision=False)
        _ = upscaler2.upscale([dummy_frame])

        # 3. CPU, fp16
        upscaler3 = ImageUpscaler(model_path="/dummy/model.pth", half_precision=True)
        _ = upscaler3.upscale([dummy_frame])

        # AFTER FIX: Should have 3 cache entries
        self.assertEqual(len(realesrgan_upscaler._MODEL_CACHE), 3)
        self.assertEqual(mock_module_to.call_count, 3)


if __name__ == "__main__":
    unittest.main()
```

```
import ast
import unittest
from pathlib import Path


class TestRealESRGANHalfPrecisionSafety(unittest.TestCase):
    def setUp(self):
        # Fallback for here-doc execution (no __file__)
        multimodal_gen_dir = Path.cwd()
        self.src_path = (
            multimodal_gen_dir
            / "runtime"
            / "postprocess"
            / "realesrgan_upscaler.py"
        )
        self.tree = ast.parse(self.src_path.read_text(encoding="utf-8"))

    def _get_method(self, class_name: str, method_name: str) -> ast.FunctionDef:
        for node in self.tree.body:
            if isinstance(node, ast.ClassDef) and node.name == class_name:
                for item in node.body:
                    if isinstance(item, ast.FunctionDef) and item.name == method_name:
                        return item
        raise AssertionError(
            f"Could not find {class_name}.{method_name} in {self.src_path}"
        )

    def test_upscale_casts_input_to_model_dtype(self):
        """Regression test: fp16 weights require fp16 inputs (no float/half mismatch)."""
        upscale_fn = self._get_method("UpscalerModel", "upscale")

        found_dtype_kw = False
        for node in ast.walk(upscale_fn):
            if not (
                isinstance(node, ast.Call)
                and isinstance(node.func, ast.Attribute)
                and node.func.attr == "to"
            ):
                continue

            dtype_kw = None
            for kw in node.keywords:
                if kw.arg == "dtype":
                    dtype_kw = kw.value
                    break
            if dtype_kw is None:
                continue

            if (
                isinstance(dtype_kw, ast.Attribute)
                and isinstance(dtype_kw.value, ast.Name)
                and dtype_kw.value.id == "self"
                and dtype_kw.attr == "dtype"
            ):
                found_dtype_kw = True
                break

        self.assertTrue(
            found_dtype_kw,
            "Expected UpscalerModel.upscale to cast input tensor dtype to match model dtype.",
        )

    def test_model_cache_key_includes_dtype(self):
        """Regression test: fp16/fp32 loads must not collide in the global cache."""
        ensure_fn = self._get_method("ImageUpscaler", "_ensure_model_loaded")

        found_cache_key = False
        for node in ast.walk(ensure_fn):
            if not isinstance(node, ast.Assign):
                continue
            if len(node.targets) != 1 or not isinstance(node.targets[0], ast.Name):
                continue
            if node.targets[0].id != "cache_key":
                continue
            if not isinstance(node.value, ast.Tuple) or len(node.value.elts) != 3:
                continue

            third = node.value.elts[2]
            if isinstance(third, ast.Name) and third.id == "target_dtype":
                found_cache_key = True
                break

        self.assertTrue(
            found_cache_key,
            "Expected ImageUpscaler._ensure_model_loaded cache_key to include target_dtype.",
        )
```
